lib/fluent/plugin/kubernetes_metadata_watch_pods.rb

```diff
@@ -24,28 +24,44 @@ module WatchPods
     include ::KubernetesMetadata::Common
 
     def start_pod_watch
-      begin
-        options = {
-          resource_version: '0'  # Fetch from API server.
-        }
-        if ENV['K8S_NODE_NAME']
-          options[:field_selector] = 'spec.nodeName=' + ENV['K8S_NODE_NAME']
-        end
-        pods = @client.get_pods(options)
-        pods.each do |pod|
-          cache_key = pod.metadata['uid']
-          @cache[cache_key] = parse_pod_metadata(pod)
-          @stats.bump(:pod_cache_host_updates)
-        end
-        options[:resource_version] = pods.resourceVersion
-        watcher = @client.watch_pods(options)
-      rescue Exception => e
-        message = "Exception encountered fetching metadata from Kubernetes API endpoint: #{e.message}"
-        message += " (#{e.response})" if e.respond_to?(:response)
+      get_pods_and_start_watcher
+    rescue Exception => e
+      message = "start_pod_watch: Exception encountered setting up pod watch " \
+                "from Kubernetes API #{@apiVersion} endpoint " \
+                "#{@kubernetes_url}: #{e.message}"
+      message += " (#{e.response})" if e.respond_to?(:response)
+      log.debug(message)
+
+      raise Fluent::ConfigError, message
+    end
 
-        raise Fluent::ConfigError, message
+    # List all pods, record the resourceVersion and return a watcher starting
+    # from that resourceVersion.
+    def get_pods_and_start_watcher
+      options = {
```

Comment:
I realize we are moving the code but why would we start from resourceVersion 0 instead of applying the same pattern that is used by getting namespaces:

1. list pods
1. get resourceversion
1. watch pods from resourceversion

Does 'watching from version 0' get every add/update/delete that exists instead of 'watch from now'?

Reply:
@jcantrill - The current implementation seems to follow your suggestion:

1. list pods (line 47)
2. get resourceversion (line 53)
3. watch pods from resourceversion (line 54)

Only the initial `get_pods` is from `resourceVersion=0`, which was recommended by the Kubernetes team because that forced it to use apiserver's local cache, which is faster.

Reply:
@qingling128 I defer to what you were advised but this is different from the namespace watch:
```    def get_namespaces_and_start_watcher
      resource_version = @client.get_namespaces.resourceVersion
      watcher = @client.watch_namespaces(resource_version)
      Thread.current[:namespace_watcher_alive] = true
      watcher
    end```

Reply:
That's probably because I didn't touch the namespaces file when I added the rv=0 logic to the pod watcher, as we hadn't noticed problems with it. It makes sense to do this in both since we're changing the code substantially.